features/cellular/TARGET_GENERIC_MODEM/generic_modem_driver/ReferenceCellularInterface.h

```diff
@@ -0,0 +1,271 @@
+/* Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef REFERENCE_CELLULAR_INTERFACE_
+#define REFERENCE_CELLULAR_INTERFACE_
+
+#include "mbed.h"
+#include "nsapi.h"
+#include "rtos.h"
+#include "FileHandle.h"
+#include "InterruptIn.h"
+#include "ATParser.h"
+#include "PinNames.h"
+
+// Forward declaration
+class NetworkStack;
+
+/**
+ * Network registration status
+ * UBX-13001820 - AT Commands Example Application Note (Section 4.1.4.5)
+ */
+typedef enum {
+   GSM=0,
+   COMPACT_GSM=1,
+   UTRAN=2,
+   EDGE=3,
+   HSDPA=4,
+   HSUPA=5,
+   HSDPA_HSUPA=6,
+   LTE=7
+} radio_access_nwk_type;
+
+/**
+ * Used in registration process to tell which type of network
+ * to connect.
+ */
+typedef enum {
+    CIRCUIT_SWITCHED=0,
+    PACKET_SWITCHED
+} nwk_type;
+
+/**
+ * Circuit Switched network registration status (CREG Usage)
+ * UBX-13001820 - AT Commands Example Application Note (Section 7.10.3)
+ */
+typedef enum {
+    CSD_NOT_REGISTERED_NOT_SEARCHING=0,
+    CSD_REGISTERED=1,
+    CSD_NOT_REGISTERED_SEARCHING=2,
+    CSD_REGISTRATION_DENIED=3,
+    CSD_UNKNOWN_COVERAGE=4,
+    CSD_REGISTERED_ROAMING=5,
+    CSD_SMS_ONLY=6,
+    CSD_SMS_ONLY_ROAMING=7,
+    CSD_CSFB_NOT_PREFERRED=9
+} nwk_registration_status_csd;
+
+/**
+ * Packet Switched network registration status (CGREG Usage)
+ * UBX-13001820 - AT Commands Example Application Note (Section 18.27.3)
+ */
+typedef enum {
+    PSD_NOT_REGISTERED_NOT_SEARCHING=0,
+    PSD_REGISTERED=1,
+    PSD_NOT_REGISTERED_SEARCHING=2,
+    PSD_REGISTRATION_DENIED=3,
+    PSD_UNKNOWN_COVERAGE=4,
+    PSD_REGISTERED_ROAMING=5,
+    PSD_EMERGENCY_SERVICES_ONLY=8
+} nwk_registration_status_psd;
+
+typedef struct {
+    char ccid[20+1];    //!< Integrated Circuit Card ID
+    char imsi[15+1];    //!< International Mobile Station Identity
+    char imei[15+1];    //!< International Mobile Equipment Identity
+    char meid[18+1];    //!< Mobile Equipment IDentifier
+    int flags;
+    bool ppp_connection_up;
+    radio_access_nwk_type rat;
+    nwk_registration_status_csd reg_status_csd;
+    nwk_registration_status_psd reg_status_psd;
+} device_info;
+
+/** ReferenceCellularInterface class
+ *
+ *  This interface serves as the controller/driver for the Cellular
+ *  modems (tested with UBLOX_C027 and MTS_DRAGONFLY_F411RE).
+ */
+class ReferenceCellularInterface : public CellularInterface {
```

Comment:
I don't believe anyone is using the existing CellularInterface (apart from demos which are also broken by the ATParser in-treeing), so I'd be fine with shuffling.

One thing I am aware of is that we may want to separate the concepts of the cellular bits of it from the networking bits. Non-networking cellular interfaces that can, for example, just send texts are a possibility. So there may be a need for further subclassing in future.

Keeping that in mind, maybe CellularInterfaceBase would be the base of a cellular network interface, and that could in future be a specialisation of a CellularBase (which would declare base stuff like PIN functions).